File: src/main/java/model/player/PlayerComputer.java
package model.player;

import model.GameData;
import model.moves.Move;
import model.util.Colour;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Consumer;

//TODO Upgrade algorithm to min/max with alpha-beta pruning
//TODO Write tests because behaves weird
//TODO consider moving pieces values (ex. pawn = 1) to this class since pieces should not be responsible for the algorithm

/**
 * Un player qui utilise un algorithm pour trouver le prochain meilleur moves.
 * L'algorithme est un algorithm recursif min-max qui utilise les valeurs des pièces (chaque pièce à une sequenceValue)
 *
 * Il existe deux niveaux de difficulté (facile et difficile)
 */
public class PlayerComputer extends Player {
    //Les niveaux de difficultés
    public static final Difficulty EASY = new Difficulty(3, "Facile");
    public static final Difficulty HARD = new Difficulty(4, "Difficile");

    /**
     * Le niveau de difficulty du player
     */
    private final Difficulty difficulty;

    /**
     * Le gamewindow data
     */
    @Nullable
    private GameData gameData;

    /**
     * @param difficulty la difficulté de l'algorithme
     */
    public PlayerComputer(Difficulty difficulty) {
        this.difficulty = difficulty;
    }

    @Override
    public void initializeGameData(@NotNull GameData gameData) {
        this.gameData = gameData;
    }

    /**
     * Retourne le moves qui retourne le plus de points
     *
     * @param callback la méthode par laquelle l'on soumet son prochain moves
     */
    @Override
    public void getMove(Consumer<Move> callback, Colour colour) {
        new Thread(() ->
                callback.accept(
                        calculateBestMove(new MoveSequence(), colour)
                                .getFirstMove()
                )
        ).start();
    }

    @NotNull
    private MoveSequence calculateBestMove(MoveSequence pastSequence, Colour colour) {
        //Si on est déjà à la profondeur maximale retourner
        if (pastSequence.getLength() == difficulty.searchDepth) return pastSequence;

        //Calculer les moves possibles
        Collection<Move> possibleMoves = gameData.getAllLegalMoves(colour);

        MoveSequence bestMove = null;

        for (Move move : possibleMoves) {
            move.apply(gameData); //Appliquer le moves

            //Calculer la sequenceValue du moves (partie récursive)
            MoveSequence moveSequence = calculateBestMove(new MoveSequence(pastSequence, move), getOppositeColour(colour));

            //Si le moves est meilleur que bestMove remplacer bestMove
            //Si le moves à la même sequenceValue, remplacer 50% du temps
            if (bestMove == null) {
                bestMove = moveSequence;
            } else {
                if (colour == Colour.WHITE) {
                    if (moveSequence.getSequenceValue() > bestMove.getSequenceValue()
                            || (moveSequence.getSequenceValue() == bestMove.getSequenceValue() && Math.random() > 0.5)) {
                        bestMove = moveSequence;
                    }
                } else {
                    if (moveSequence.getSequenceValue() < bestMove.getSequenceValue()
                            || (moveSequence.getSequenceValue() == bestMove.getSequenceValue() && Math.random() > 0.5)) {
                        bestMove = moveSequence;
                    }
                }
            }

            move.undo(gameData); //Défaire les changements
        }

        return bestMove == null ? pastSequence : bestMove;
    }

    /**
     * Si blanc retourne noir et vice-versa
     *
     * @param colour la colour à getOppositeColour
     * @return la colour inversée
     */
    private Colour getOppositeColour(Colour colour) {
        return colour == Colour.WHITE ? Colour.BLACK : Colour.WHITE;
    }

    /**
     * Une classe qui représente une série de moves et la sequenceValue de la série
     */
    private class MoveSequence {
        private final List<Move> moves;

        /**
         * La somme de la sequenceValue de toutes les pièces
         */
        private final int sequenceValue;

        MoveSequence() {
            this.sequenceValue = 0;
            this.moves = new ArrayList<>();
        }

        private MoveSequence(MoveSequence moveSequence, Move move) {
            this.moves = new ArrayList<>(moveSequence.moves);
            this.moves.add(move);
            this.sequenceValue = moveSequence.sequenceValue + move.getValue();
        }

        int getSequenceValue() {
            return sequenceValue;
        }

        int getLength() {
            return moves.size();
        }

        Move getFirstMove() {
            return moves.get(0);
        }
    }

    /**
     * La difficulté de l'algorithm (dépendant de la profondeur)
     */
    private static class Difficulty implements Serializable {
        private final int searchDepth;
        private final String name;

        Difficulty(int profondeur, String name) {
            this.searchDepth = profondeur;
            this.name = name;
        }
    }

    @Override
    public String getName() {
        return "Ordinateur (" + difficulty.name + ")";
    }
}

File: src/main/java/model/player/PlayerComputer.kt
package model.player

import model.GameData
import model.moves.Move
import model.util.Colour
import java.io.Serializable
import java.util.*
import java.util.function.Consumer

//TODO Upgrade algorithm to min/max with alpha-beta pruning
//TODO Write tests because behaves weird
//TODO consider moving pieces values (ex. pawn = 1) to this class since pieces should not be responsible for this algorithm

/**
 * This player uses an algorithm to find the best next move.
 * The algorithm is a recursive min-max algorithm that uses the relative value of each move to evaluate the best move
 *
 *
 * There are two difficulty levels (Easy and Hard) they change the search depth of the algorithm
 */
class PlayerComputer
/**
 * @param difficulty the difficulty level for the algorithm
 */
(
        /**
         * The difficulty level of the current instance of this player
         */
        private val difficulty: Difficulty) : Player() {

    /**
     * The game data
     */
    private var gameData: GameData? = null

    override val name: String = "Computer (" + difficulty.name + ")"

    override fun initializeGameData(gameData: GameData) {
        this.gameData = gameData
    }

    /**
     * Calculates the best move and returns it via the callback.
     */
    override fun getMove(callback: Consumer<Move>, colour: Colour) {
        Thread {
            callback.accept(
                    calculateBestMove(MoveSequence(), colour)
                            .firstMove
            )
        }.start()
    }

    private fun calculateBestMove(pastSequence: MoveSequence, colour: Colour): MoveSequence {
        //If we've reached the max depth return the current sequence
        if (pastSequence.length == difficulty.searchDepth) return pastSequence

        //Calculate all the possible moves
        val possibleMoves = gameData!!.getPossibleLegalMoves(colour)

        //The best move
        var bestMove: MoveSequence? = null

        for (move in possibleMoves) {
            move.apply(gameData!!) //Apply the move to the data

            //Calculate the value of this move (using recursion)
            val moveSequence = calculateBestMove(MoveSequence(pastSequence, move), getOppositeColour(colour))

            //If the move is better than bestMove update bestMove.
            //If the move is equal to bestMove update bestMove 50% of the time (to allow for variation)
            if (bestMove == null) {
                bestMove = moveSequence
            } else {
                if (colour == Colour.WHITE) {
                    if (moveSequence.sequenceValue > bestMove.sequenceValue || moveSequence.sequenceValue == bestMove.sequenceValue && Math.random() > 0.5) {
                        bestMove = moveSequence
                    }
                } else {
                    if (moveSequence.sequenceValue < bestMove.sequenceValue || moveSequence.sequenceValue == bestMove.sequenceValue && Math.random() > 0.5) {
                        bestMove = moveSequence
                    }
                }
            }

            move.undo(gameData!!) //Undo changes
        }

        return if (bestMove == null) pastSequence else bestMove
    }

    /**
     * @return the opposite colour (used to switch turns)
     */
    private fun getOppositeColour(colour: Colour): Colour {
        return if (colour == Colour.WHITE) Colour.BLACK else Colour.WHITE
    }

    /**
     * An object representing a series of moves and its value
     */
    private inner class MoveSequence {
        private val moves: LinkedList<Move>

        /**
         * The sum of the value of each move
         */
        internal val sequenceValue: Int

        internal val length: Int
            get() = moves.size

        internal val firstMove: Move
            get() = moves.first

        internal constructor() {
            this.sequenceValue = 0
            this.moves = LinkedList()
        }

        internal constructor(moveSequence: MoveSequence, move: Move) {
            this.moves = LinkedList(moveSequence.moves)
            this.moves.add(move)
            this.sequenceValue = moveSequence.sequenceValue + move.value
        }

        override fun toString(): String {
            val stringBuilder = StringBuilder()

            moves.forEach { obj -> stringBuilder.append(obj).append(" ") }

            return stringBuilder.toString()
        }
    }

    /**
     * The difficulty of the algorithm. Has a search depth and a name
     */
    class Difficulty internal constructor(internal val searchDepth: Int, internal val name: String) : Serializable

    companion object {
        //The difficulty levels
        val EASY = Difficulty(3, "Easy")
        val HARD = Difficulty(4, "Hard")
    }
}